pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,288 @@
+		o.k8sClient = client.NewDryRunClient(o.k8sClient)
+	}
+
+	return nil
+}
+
+func runE(cmd *cobra.Command, _ []string) error {
+	if err := opts.complete(cmd); err != nil {
+		return err
+	}
+
+	writer := cmd.ErrOrStderr()
+	crdNamesToUpgrade, err := getCRDNamesToUpgrade(writer, opts.k8sClient, sets.New[string](opts.crdNames...))
+	if err != nil {
+		return err
+	}
+
+	crdObjectsToUpgrade := getCRDsToUpgrade(writer, opts.k8sClient, crdNamesToUpgrade)
+	for _, crd := range crdObjectsToUpgrade {
+		if err = upgradeCRDObjects(writer, opts.k8sClient, crd); err != nil {
+			fmt.Fprintf(writer, "Failed to upgrade CRD %q: %v.\n", crd.Name, err)
+		}
+		if err = patchCRDStoredVersions(opts.k8sClient, crd); err != nil {
+			fmt.Fprintf(writer, "Failed to patch CRD %q status.storedVersion: %v.\n", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getAntreaCRDNames gets names of all Antrea CRDs.
+func getAntreaCRDNames(k8sClient client.Client) (sets.Set[string], error) {
+	crdNames := sets.New[string]()
+	crdList := &apiextv1.CustomResourceDefinitionList{}
+	if err := k8sClient.List(context.TODO(), crdList); err != nil {
+		return nil, fmt.Errorf("failed to get CRD list: %v", err)
+	}
+	for _, crd := range crdList.Items {
+		if strings.HasSuffix(crd.Name, "crd.antrea.io") {
+			crdNames.Insert(crd.Name)
+		}
+	}
+
+	return crdNames, nil
+}
+
+// getCRDNamesToUpgrade gets names of Antrea CRDs to upgrade.
+func getCRDNamesToUpgrade(writer io.Writer, k8sClient client.Client, crdNamesToUpgrade sets.Set[string]) (sets.Set[string], error) {
+	antreaCRDNames, err := getAntreaCRDNames(k8sClient)
+	if err != nil {
+		return nil, err
+	}
+	// If the user-provided name list of CRDs to upgrade is empty, upgrade all Antrea CRDs.
+	if crdNamesToUpgrade.Len() == 0 {
+		crdNamesToUpgrade = antreaCRDNames
+	} else {
+		// If the user-provided name list of CRDs to upgrade is not empty, and it contains CRDs without suffix
+		// "crd.antrea.io", skip these CRDs.
+		for name := range crdNamesToUpgrade.Difference(antreaCRDNames) {
+			fmt.Fprintf(writer, "Skip CRD %q which is not created by Antrea.\n", name)
+		}
+		// Only upgrade the CRDs with suffix "crd.antrea.io".
+		crdNamesToUpgrade = crdNamesToUpgrade.Intersection(antreaCRDNames)
+	}
+	return crdNamesToUpgrade, nil
+}
+
+// getCRDsToUpgrade gets a list of Antrea CRDs to upgrade.
+func getCRDsToUpgrade(writer io.Writer, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(writer, "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) == 1 {
+			fmt.Fprintf(writer, "Skip upgrading CRD %q since it only has one version.\n", crdName)
```

Comment:
I think this is an info instead of an error, you may pass cmd.OutOrStdout() for this kind of information. Please also check other logs.